wom/src/main/scala/wom/types/WomType.scala

```diff
@@ -98,10 +98,112 @@ object WomType {
     }
   }
 
-  def lowestCommonSubtype(types: Iterable[WomType]): WomType = {
-    types.collectFirst {
-      case t1 if types.forall(t2 => t1.isCoerceableFrom(t2)) => t1
-    } getOrElse WomAnyType
+  def lowestCommonSubtype(types: Iterable[WomType]): WomType = types match {
+    case e if e.isEmpty => WomNothingType
+    case ListOfPrimitives(primitiveType) => primitiveType
+    case ListOfPairs(pairType) => pairType
+    case ListOfOptionals(optionalType) => optionalType
+    case ListOfMaps(mapType) => mapType
+    case ListOfArrays(arrayType) => arrayType
+    case ListOfObject(objectType) => objectType
+    case _ => WomAnyType
+  }
+
+  private object ListOfPrimitives {
+    def unapply(types: Iterable[WomType]): Option[WomType] = {
+      if (types.forall(_.isInstanceOf[WomPrimitiveType])) {
+        firstCommonPrimitive(types)
+      } else None
+    }
+
+    val coercePriority = List(
+      WomStringType, WomSingleFileType, WomUnlistedDirectoryType, WomFloatType, WomIntegerType, WomBooleanType, WomObjectType
+    )
+
+    private def firstCommonPrimitive(types: Iterable[WomType]): Option[WomType] = {
+      // Types in the incoming list which everything could coerce to:
+      val suppliedOptions = types.filter(t => types.forall(t.isCoerceableFrom))
+
+      // A type not in the incoming list but which everything could coerce to nonetheless:
+      lazy val unsuppliedOption: Option[WomType] = coercePriority.find(p => types.forall(p.isCoerceableFrom))
+
+      coercePriority.find { p => suppliedOptions.toList.contains(p) } orElse { unsuppliedOption }
+    }
+  }
+
+
+  private object ListOfPairs {
+    def unapply(types: Iterable[WomType]): Option[WomPairType] = {
+      if (types.forall(_.isInstanceOf[WomPairType])) {
+        val pairs = types.map(_.asInstanceOf[WomPairType])
+        val leftType = lowestCommonSubtype(pairs.map(_.leftType))
+        val rightType = lowestCommonSubtype(pairs.map(_.rightType))
+        Some(WomPairType(leftType, rightType))
+      } else None
+    }
+  }
+
+  private object ListOfOptionals {
+    def unapply(types: Iterable[WomType]): Option[WomOptionalType] = {
+      val atLeastOneOptional = types.exists {
```

Comment:
This could be shorter and (imo) clearer as `types.exists(_.isInstanceOf[WomOptionalType])`

Reply:
Yes, in this case... the reason for doing this at all is that some types have custom `unapply` methods (eg `WomArray` which also scoops up `WomMap`s as arrays of pairs).

But perhaps it's best to make that explicit rather than forcing everything else through the same weird pattern